programs/clearing_house/src/math/margin.rs

```diff
@@ -1,91 +1,260 @@
+        return Ok(asset_weight);
+    }
+
+    let size_sqrt = (size / 1000).nth_root(2) + 1; //1e13 -> 1e10 -> 1e5
+    let imf_numerator = BANK_IMF_PRECISION + BANK_IMF_PRECISION / 10;
+
+    let size_discount_asset_weight = imf_numerator
+        .checked_mul(BANK_WEIGHT_PRECISION)
+        .ok_or_else(math_error!())?
+        .checked_div(
+            BANK_IMF_PRECISION
+                .checked_add(
+                    size_sqrt // 1e5
+                        .checked_mul(imf_factor)
+                        .ok_or_else(math_error!())?
+                        .checked_div(100_000) // 1e5
+                        .ok_or_else(math_error!())?,
+                )
+                .ok_or_else(math_error!())?,
+        )
+        .ok_or_else(math_error!())?;
+
+    let min_asset_weight = min(asset_weight, size_discount_asset_weight);
+
+    Ok(min_asset_weight)
+}
+
+pub fn calculate_bank_equity_value(
+    user_bank_balance: &UserBankBalance,
+    bank: &Bank,
+    oracle_price_data: &OraclePriceData,
+    margin_requirement_type: MarginRequirementType,
+) -> ClearingHouseResult<u128> {
+    let (balance_value, token_amount) =
+        get_balance_value_and_token_amount(user_bank_balance, bank, oracle_price_data)?;
+
+    let balance_equity_value = match user_bank_balance.balance_type {
+        BankBalanceType::Deposit => balance_value
+            .checked_mul(bank.get_asset_weight(token_amount, &margin_requirement_type)?)
+            .ok_or_else(math_error!())?
+            .checked_div(BANK_WEIGHT_PRECISION)
+            .ok_or_else(math_error!())?,
+        BankBalanceType::Borrow => balance_value
+            .checked_mul(bank.get_liability_weight(token_amount, &margin_requirement_type)?)
+            .ok_or_else(math_error!())?
+            .checked_div(BANK_WEIGHT_PRECISION)
+            .ok_or_else(math_error!())?,
+    };
+
+    Ok(balance_equity_value)
+}
+
+pub fn calculate_oracle_price_for_margin(
+    market_position: &MarketPosition,
+    market: &Market,
+    oracle_price_data: &OraclePriceData,
+) -> ClearingHouseResult<i128> {
+    let oracle_price_offset = min(
+        (market.amm.max_spread as i128)
+            .checked_mul(oracle_price_data.price)
+            .ok_or_else(math_error!())?
+            .checked_div(BID_ASK_SPREAD_PRECISION_I128)
+            .ok_or_else(math_error!())?,
+        cast_to_i128(
+            oracle_price_data
+                .confidence
+                .checked_add(market.amm.base_spread as u128)
+                .ok_or_else(math_error!())?,
+        )?,
+    );
+    let oracle_price = if market_position.base_asset_amount > 0 {
+        oracle_price_data
+            .price
+            .checked_sub(oracle_price_offset)
+            .ok_or_else(math_error!())?
+    } else {
+        oracle_price_data
+            .price
+            .checked_add(oracle_price_offset)
+            .ok_or_else(math_error!())?
+    };
+
+    Ok(oracle_price)
+}
+
+pub fn calculate_perp_equity_value(
```

Comment:
also dont know about calling this `equity`

Reply:
`calculate_perp_position_value`